File: openhpi/trunk/baselibs/python/openhpi_baselib/OhpiUtil.py

#      -*- python -*-
#
# Copyright (C) 2012, Pigeon Point Systems
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  This
# file and program are licensed under a BSD style license.  See
# the Copying file included with the OpenHPI distribution for
# full licensing terms.
#
# Author(s):
#      Anton Pak <anton.pak@pigeonpoint.com>
#

from openhpi_baselib.HpiDataTypes import *
from openhpi_baselib.OhpiDataTypes import *


#**********************************************************
# OHPI Utility Functions
#**********************************************************

#**********************************************************
# NB: toXXX raises ValueError if lookup fails
#**********************************************************

#**********************************************************
# Check Functions for OHPI Complex Data Types
#**********************************************************

#**
# Check function for OHPI struct oHpiHandlerConfigT
#**
def checkoHpiHandlerConfigT( x ):
    if x is None:
        return False
    if not isinstance( x, oHpiHandlerConfigT ):
        return False
    if x.items is None:
        return False
    if not isinstance( x.items, list ):
        return False
    for nv in x.items:
        if not isinstance( nv, tuple ):
            return False
        if len( nv ) != 2:
            return False
        ( name, value ) = nv
        if not isinstance( name, str ):
            return False
        if len( name ) != SAHPI_MAX_TEXT_BUFFER_LENGTH:
            return False
        if not isinstance( value, str ):
            return False
        if len( value ) != SAHPI_MAX_TEXT_BUFFER_LENGTH:
            return False
    return True


#**********************************************************
# oHpiHandlerConfigT Helpers
#**********************************************************

def fromoHpiHandlerConfigT( config ):
    d = dict()
    for ( n, v ) in config.items:
        d[n.rstrip( chr(0) )] = v.rstrip( chr(0) )
    return d

def tooHpiHandlerConfigT( d ):
    if not isinstance( d, dict ):
        raise ValueError()
    if len( d ) > 255:
        raise ValueError()
    for ( n, v ) in d.iteritems():
        if not isinstance( n, str ):
            raise ValueError()
        if len( n ) > SAHPI_MAX_TEXT_BUFFER_LENGTH:
            raise ValueError()
        if not isinstance( v, str ):
            raise ValueError()
        if len( v ) > SAHPI_MAX_TEXT_BUFFER_LENGTH:
            raise ValueError()
    config = oHpiHandlerConfigT()
    config.items = []
    for ( n, v ) in d.iteritems():
        n = n.ljust( SAHPI_MAX_TEXT_BUFFER_LENGTH, chr(0) )
        v = v.ljust( SAHPI_MAX_TEXT_BUFFER_LENGTH, chr(0) )
        config.items.append( ( n, v ) )
    return config

File: openhpi/trunk/baselibs/java/openhpi_baselib/OhpiUtil.java
/*      -*- java -*-
 *
 * Copyright (C) 2012, Pigeon Point Systems
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  This
 * file and program are licensed under a BSD style license.  See
 * the Copying file included with the OpenHPI distribution for
 * full licensing terms.
 *
 * Author(s):
 *      Anton Pak <anton.pak@pigeonpoint.com>
 */

package org.openhpi;

import java.io.UnsupportedEncodingException;
import java.lang.IllegalArgumentException;
import java.lang.Math;
import java.util.AbstractMap;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Map.Entry;

import static org.openhpi.HpiDataTypes.*;
import static org.openhpi.OhpiDataTypes.*;


public class OhpiUtil
{
    /**********************************************************
     * NB: ToXXX throws IllegalArgumentException if lookup fails
     *********************************************************/

    // Just to ensure nobody creates it
    private OhpiUtil()
    {
        // empty
    }


    /**********************************************************
     * Check Functions for OHPI Complex Data Types
     *********************************************************/

    /**
     * Check function for OHPI struct oHpiHandlerConfigT
     */
    public static boolean check( oHpiHandlerConfigT x )
    {
        if ( x == null ) {
            return false;
        }
        if ( x.items == null ) {
            return false;
        }
        for ( Entry<byte[], byte[]> kv : x.items ) {
            byte[] k = kv.getKey();
            if ( k == null ) {
                return false;
            }
            if ( k.length != SAHPI_MAX_TEXT_BUFFER_LENGTH ) {
                return false;
            }
            byte[] v = kv.getValue();
            if ( v == null ) {
                return false;
            }
            if ( v.length != SAHPI_MAX_TEXT_BUFFER_LENGTH ) {
                return false;
            }
        }
        return true;
    }


    /**********************************************************
     * oHpiHandlerConfigT Helpers
     *********************************************************/
    public static Map<String, String> fromoHpiHandlerConfigT( oHpiHandlerConfigT config )
    {
        HashMap<String, String> hm = new HashMap<String, String>();
        for ( Entry<byte[], byte[]> kv : config.items ) {
            try {
                hm.put( new String( kv.getKey(), "US-ASCII" ),
                        new String( kv.getValue(), "US-ASCII" ) );
            } catch ( UnsupportedEncodingException e ) {
                throw new IllegalArgumentException();
            }
        }

        return hm;
    }

    public static oHpiHandlerConfigT tooHpiHandlerConfigT( Map<String, String> hm )
    {
        oHpiHandlerConfigT config = new oHpiHandlerConfigT();
        config.items = new LinkedList< Entry<byte[], byte[]> >();
        if ( hm.size() > 255 ) {
            throw new IllegalArgumentException();
        }
        for ( Entry<String, String> kv : hm.entrySet() ) {
            int len;
            byte[] encoded;

            String k = kv.getKey();
            if ( k == null ) {
                throw new IllegalArgumentException();
            }
            byte[] name = new byte[SAHPI_MAX_TEXT_BUFFER_LENGTH];
            try {
                encoded = k.getBytes( "US-ASCII" );
            } catch ( UnsupportedEncodingException e ) {
                throw new IllegalArgumentException();
            }
            len = Math.min( encoded.length, SAHPI_MAX_TEXT_BUFFER_LENGTH );
            System.arraycopy( encoded, 0, name, 0, len );

            String v = kv.getValue();
            if ( v == null ) {
                throw new IllegalArgumentException();
            }
            byte[] val = new byte[SAHPI_MAX_TEXT_BUFFER_LENGTH];
            try {
                encoded = v.getBytes( "US-ASCII" );
            } catch ( UnsupportedEncodingException e ) {
                throw new IllegalArgumentException();
            }
            len = Math.min( encoded.length, SAHPI_MAX_TEXT_BUFFER_LENGTH );
            System.arraycopy( encoded, 0, val, 0, len );

            config.items.add( new AbstractMap.SimpleEntry<byte[], byte[]>( name, val ) );
        }

        return config;
    }


};

